10gen/envoy-serverless
# CLOUDP-134864: Envoy parses and sends proxy protocol TLV to upstream as HTTP headers

[CLOUDP-134864](https://jira.mongodb.org/browse/CLOUDP-134864)

## Changes 
- Creates a new substitution formatter to parse the proxy protocol TLV values stored in the filter state and add them as an HTTP header to be used upstream (usage: `%PROXY_PROTOCOL_TLVS(<tlv_type>)%`). Note that I took the approach of adding a substitution formatter instead of a header formatter because of the deprecation of legacy header formatters noted at the [bottom of this Envoy docs page](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#custom-request-response-headers). (they're very similar though) 

## QA 

### Manual testing notes 
I put up a patch of the adjustments I made to the client/server proxy protocol POC in the serverless-proxy repo to be able to manually test this PR. Feel free to check out my branch if you'd like to do your own manual testing. 
[Patch](https://github.com/10gen/serverless-proxy/pull/32)
[Wiki page with instructions to run the POC](https://wiki.corp.mongodb.com/display/MMS/Envoy+POC+of+Proxy+Protocol)

#### Example output from running the POC
Envoy log lines indicating that the expected HTTP header was added correctly and received upstream: 
```
[2023-01-11 22:23:45.525][4143][debug][router] [source/common/router/router.cc:678] [C16][S14622910830797364673] router decoding headers:
':authority', 'pong.foo.bar'
':path', '/'
':method', 'POST'
':scheme', 'https'
'content-length', '4'
'x-forwarded-proto', 'https'
'x-request-id', 'e1fc4ecb-7319-4bc3-9d80-a609366a229e'
'x-envoy-expected-rq-timeout-ms', '15000'
'proxy-protocol-tlv', 'AAAAAAAAAAAAAAAAAAAAAQ=='
```

Client log lines from manual testing with the proxy protocol POC:
```
DEBUG:root:Connecting to ('pong.foobar.com', 6666)
INFO:root:Sending header: family: 2, protocol: 1, source: 1.2.3.4:6666, dest: 4.3.2.1:6666, tlvs: [<type: b'\x05', uuid: 00000000-0000-0000-0000-000000000001>]
DEBUG:root:Sending header of length 47: b'\r\n\r\n\x00\r\nQUIT\n!\x11\x00\x1f\x01\x02\x03\x04\x04\x03\x02\x01\x1a\n\x1a\n\x05\x00\x10\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01'
INFO:root:Sending PING
INFO:root:Received HTTP/1.1 200 OK
server: envoy
content-length: 4
content-type: text/plain
x-envoy-upstream-service-time: 1
date: Wed, 18 Jan 2023 20:15:36 GMT

PONG
```

Server log lines from manual testing with the proxy protocol POC:
 ```
DEBUG:root:Received connection from ('127.0.0.1', 39288)
INFO:root:Noop handler, skipping proxy protocol header...
INFO:root:Received http data: b'POST / HTTP/1.1\r\nhost: pong.foo.bar\r\ncontent-length: 4\r\nx-forwarded-proto: https\r\nx-request-id: 4b642052-38bc-46f9-bdbe-d166c5a04547\r\nx-envoy-expected-rq-timeout-ms: 15000\r\nproxy-protocol-tlv: AAAAAAAAAAAAAAAAAAAAAQ==\r\n\r\n'
INFO:root:Received http body: b'PING'
INFO:root:Sending: PONG
```
